File: MeshChatMVP/AlertTrustEngine.swift
import Foundation

// MARK: - Output types

struct ScoredAlert {
    let alert: Alert
    let score: Double
}

struct AlertCluster: Identifiable {
    let id = UUID()
    let alerts: [ScoredAlert]

    var clusterScore: Double { alerts.reduce(0) { $0 + $1.score } }
    var leadAlert: Alert { alerts.max(by: { $0.score < $1.score })!.alert }
}

// MARK: - Engine

enum AlertTrustEngine {

    // MARK: Configurable weights
    static let wAuthor: Double = 10
    static let wVouch:  Double = 5
    static let wDeny:   Double = 10

    // MARK: Relationship multipliers
    static let rFriend:    Double = 1.0
    static let rAssociate: Double = 0.5
    static let rStranger:  Double = 0.2

    // MARK: Proximity multipliers
    static let pClose: Double = 1.0
    static let pFar:   Double = 0.5
    static let proximityGPSThresholdMeters: Double = 500
    static let proximityRSSIThreshold: Int = -70

    // MARK: Time decay — half-life of 2 hours
    // e^(-λΔt) = 0.5 when Δt = 2 h  →  λ = ln(2) / (2 * 3600)
    static let decayLambda: Double = log(2.0) / (2.0 * 3600.0)

    // MARK: Clustering thresholds
    static let clusterRadiusMeters:      Double = 500
    static let clusterTimeWindowSeconds: Double = 15 * 60

    // MARK: - Score

    /// Computes S_alert using the formula:
    /// ((W_author·R_author) + Σ(W_vouch·R_i·P_i) − Σ(W_deny·R_j·P_j)) · e^(−λΔt)
    ///
    /// - Parameters:
    ///   - alert: The alert to score.
    ///   - vouches: All vouches recorded for this alert.
    ///   - relationships: nodeID → "friend" | "associate" (absent = stranger).
    ///     Build from `SavedContact` entries using `canonicalSenderID`.
    ///   - sightings: nodeID → most-recent `NodeSighting` for proximity lookup.
    ///   - now: Reference time; injectable for unit testing.
    static func score(
        alert: Alert,
        vouches: [Vouch],
        relationships: [String: String],
        sightings: [String: NodeSighting],
        now: Date = Date()
    ) -> Double {
        let age   = now.timeIntervalSince1970 - Double(alert.createdAt)
        let decay = exp(-decayLambda * max(0.0, age))

        let rAuthor = relationshipMultiplier(relationships[alert.authorID])
        var total   = wAuthor * rAuthor

        for vouch in vouches {
            let r = relationshipMultiplier(relationships[vouch.voucherID])
            let p = proximityMultiplier(
                alertLat: alert.lat, alertLon: alert.lon,
                sighting: sightings[vouch.voucherID]
            )
            if vouch.value > 0 {
                total += wVouch * r * p
            } else {
                total -= wDeny  * r * p
            }
        }

        return total * decay
    }

    // MARK: - Clustering

    /// Groups `ScoredAlert`s into clusters (≤500 m apart, ≤15 min apart),
    /// then returns them sorted descending by cluster score.
    static func cluster(scoredAlerts: [ScoredAlert]) -> [AlertCluster] {
        var remaining = scoredAlerts
        var clusters: [AlertCluster] = []

        while !remaining.isEmpty {
            let seed    = remaining.removeFirst()
            var members = [seed]
            var i       = 0

            while i < remaining.count {
                let candidate = remaining[i]
                let dist = haversineMeters(
                    lat1: seed.alert.lat, lon1: seed.alert.lon,
                    lat2: candidate.alert.lat, lon2: candidate.alert.lon
                )
                let timeDiff = abs(Double(seed.alert.createdAt - candidate.alert.createdAt))

                if dist <= clusterRadiusMeters && timeDiff <= clusterTimeWindowSeconds {
                    members.append(remaining.remove(at: i))
                } else {
                    i += 1
                }
            }

            clusters.append(AlertCluster(alerts: members))
        }

        return clusters.sorted { $0.clusterScore > $1.clusterScore }
    }

    // MARK: - Helpers

    private static func relationshipMultiplier(_ relationship: String?) -> Double {
        switch relationship {
        case SavedContact.friend:    return rFriend
        case SavedContact.associate: return rAssociate
        default:                     return rStranger
        }
    }

    /// P multiplier for a voucher: GPS distance to alert location first,
    /// then RSSI from latest sighting, then default to pFar.
    private static func proximityMultiplier(
        alertLat: Double, alertLon: Double,
        sighting: NodeSighting?
    ) -> Double {
        guard let s = sighting else { return pFar }
        if s.lat != 0 || s.lon != 0 {
            let dist = haversineMeters(lat1: alertLat, lon1: alertLon, lat2: s.lat, lon2: s.lon)
            if dist <= proximityGPSThresholdMeters { return pClose }
        }
        return s.rssi > proximityRSSIThreshold ? pClose : pFar
    }

    private static func haversineMeters(
        lat1: Double, lon1: Double,
        lat2: Double, lon2: Double
    ) -> Double {
        let R    = 6_371_000.0
        let toRad: (Double) -> Double = { $0 * .pi / 180 }
        let dLat = toRad(lat2 - lat1)
        let dLon = toRad(lon2 - lon1)
        let a    = sin(dLat / 2) * sin(dLat / 2)
                 + cos(toRad(lat1)) * cos(toRad(lat2)) * sin(dLon / 2) * sin(dLon / 2)
        return R * 2 * atan2(sqrt(a), sqrt(1 - a))
    }
}

File: MeshMapAndroid/MeshChatAndroid/app/src/main/java/com/meshchat/mvp/model/AlertTrustEngine.kt
package com.meshchat.mvp.model

import com.meshchat.mvp.database.NodeSightingEntity
import com.meshchat.mvp.database.SavedContactEntity
import kotlin.math.*

// ── Output types ──────────────────────────────────────────────────────────────

data class ScoredLabel(
    val payload: MapLabelPayload,
    val score: Double,
    val voteCount: Int
)

data class LabelEventCluster(
    val id: java.util.UUID = java.util.UUID.randomUUID(),
    val labels: List<ScoredLabel>
) {
    val clusterScore: Double get() = labels.sumOf { it.score }
    val leadScoredLabel: ScoredLabel get() = labels.maxByOrNull { it.score }!!
    val leadLabel: MapLabelPayload get() = leadScoredLabel.payload
}

// ── Engine ────────────────────────────────────────────────────────────────────

object AlertTrustEngine {

    // Weights
    const val W_AUTHOR = 10.0
    const val W_VOUCH  = 5.0
    const val W_DENY   = 10.0

    // Relationship multipliers
    const val R_FRIEND    = 1.0
    const val R_ASSOCIATE = 0.5
    const val R_STRANGER  = 0.2

    // Proximity multipliers
    const val P_CLOSE = 1.0
    const val P_FAR   = 0.5
    const val PROXIMITY_GPS_THRESHOLD_METERS = 500.0
    const val PROXIMITY_RSSI_THRESHOLD = -70

    // Time decay — half-life of 2 hours
    val DECAY_LAMBDA = ln(2.0) / (2.0 * 3600.0)

    // Clustering thresholds
    const val CLUSTER_RADIUS_METERS      = 500.0
    const val CLUSTER_TIME_WINDOW_SECS   = 15.0 * 60.0

    fun scoreLabel(
        authorID: String,
        lat: Double,
        lon: Double,
        timestampMs: Long,
        votes: Map<String, Int>,
        relationships: Map<String, String>,
        sightings: Map<String, NodeSightingEntity>,
        myDeviceID: String = "",
        nowMs: Long = System.currentTimeMillis()
    ): Double {
        val createdAtSec = timestampMs / 1000.0
        val ageSeconds   = (nowMs / 1000.0) - createdAtSec
        val decay        = exp(-DECAY_LAMBDA * maxOf(0.0, ageSeconds))

        val rAuthor = if (authorID == myDeviceID) R_FRIEND
                      else relationshipMultiplier(relationships[authorID])
        var total = W_AUTHOR * rAuthor

        for ((voucherID, vote) in votes) {
            val r = relationshipMultiplier(relationships[voucherID])
            val p = proximityMultiplier(lat, lon, sightings[voucherID])
            if (vote > 0) total += W_VOUCH * r * p
            else          total -= W_DENY  * r * p
        }

        return total * decay
    }

    fun clusterLabels(scoredLabels: List<ScoredLabel>): List<LabelEventCluster> {
        val remaining = scoredLabels.toMutableList()
        val clusters  = mutableListOf<LabelEventCluster>()

        while (remaining.isNotEmpty()) {
            val seed    = remaining.removeFirst()
            val members = mutableListOf(seed)
            val iter    = remaining.iterator()

            while (iter.hasNext()) {
                val candidate = iter.next()
                val dist = haversineMeters(
                    seed.payload.lat, seed.payload.lon,
                    candidate.payload.lat, candidate.payload.lon
                )
                val timeDiff = abs(
                    seed.payload.timestamp / 1000.0 - candidate.payload.timestamp / 1000.0
                )
                if (dist <= CLUSTER_RADIUS_METERS && timeDiff <= CLUSTER_TIME_WINDOW_SECS) {
                    members.add(candidate)
                    iter.remove()
                }
            }
            clusters.add(LabelEventCluster(labels = members))
        }

        return clusters.sortedByDescending { it.clusterScore }
    }

    private fun relationshipMultiplier(relationship: String?): Double = when (relationship) {
        SavedContactEntity.FRIEND    -> R_FRIEND
        SavedContactEntity.ASSOCIATE -> R_ASSOCIATE
        else                         -> R_STRANGER
    }

    private fun proximityMultiplier(
        alertLat: Double, alertLon: Double,
        sighting: NodeSightingEntity?
    ): Double {
        sighting ?: return P_FAR
        if (sighting.lat != 0.0 || sighting.lon != 0.0) {
            val dist = haversineMeters(alertLat, alertLon, sighting.lat, sighting.lon)
            if (dist <= PROXIMITY_GPS_THRESHOLD_METERS) return P_CLOSE
        }
        return if (sighting.rssi > PROXIMITY_RSSI_THRESHOLD) P_CLOSE else P_FAR
    }

    fun haversineMeters(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val R    = 6_371_000.0
        val toRad = { d: Double -> d * PI / 180 }
        val dLat = toRad(lat2 - lat1)
        val dLon = toRad(lon2 - lon1)
        val a    = sin(dLat / 2).pow(2) +
                   cos(toRad(lat1)) * cos(toRad(lat2)) * sin(dLon / 2).pow(2)
        return R * 2 * atan2(sqrt(a), sqrt(1 - a))
    }
}
